Fix Oracle SSL tests

jeff303 opened the pull request:
Define new test for Oracle SSL connectivity, in oracle_test.clj, similar to how things work in mysql_test.clj

Add new test util macro, with-env-keys-renamed-by, to support running tests with environ keys temporarily renamed

Using new test macro from both MySQL and Oracle SSL connectivity tests

Removing now unneeded be-tests-oracle-ssl-ee CircleCI job

Removing now unneeded test-selector parameter for test-driver orbv in CircleCI config.yml

jeff303 committed:
Fix Oracle SSL tests

Define new test for Oracle SSL connectivity, in oracle_test.clj, similar to how things work in mysql_test.clj

Add new test util macro, with-env-keys-renamed-by, to support running tests with environ keys temporarily renamed

Using new test macro from both MySQL and Oracle SSL connectivity tests

Removing now unneeded be-tests-oracle-ssl-ee CircleCI job

Removing now unneeded test-selector parameter for test-driver orbv in CircleCI config.yml

Updating JVM_OPTS to use a trust store that starts with cacerts and adds the RDS root CA, rather than one only
containing the RDS root CA

jeff303 commented:
Builds on top of #15172.

dpsutton commented:
couple things about your macro `with-env-keys-renamed-by`
It iterates through every key. I think you could probably just leverage `clojure.set/rename-keys`

```clojure
(tu/with-env-keys-whatever {:mb-mysql-test :mb-mysql-ssl-test} 
    (is (= ....))
```
Or you could take an arbitrary function `f` and call it on the env and redef with that value.

```clojure
(tu/with-env-keys-whatever #(set/rename-keys % {:mb-mysql-test :mb-mysql-ssl-test})
  (is (= ...)))
```

That would allow more manipulation (if desired) than just renaming keys. You could duplicate, swap, what have you. But the api to take in every key as a string and return a new string seems a bit off.

 But the function that maps over all keys just seems heavy

jeff303 commented:
> couple things about your macro `with-env-keys-renamed-by`
> It iterates through every key. I think you could probably just leverage `clojure.set/rename-keys`
> 
> ```clojure
> (tu/with-env-keys-whatever {:mb-mysql-test :mb-mysql-ssl-test} 
>     (is (= ....))
> ```
> 
> Or you could take an arbitrary function `f` and call it on the env and redef with that value.
> 
> ```clojure
> (tu/with-env-keys-whatever #(set/rename-keys % {:mb-mysql-test :mb-mysql-ssl-test})
>   (is (= ...)))
> ```
> 
> That would allow more manipulation (if desired) than just renaming keys. You could duplicate, swap, what have you. But the api to take in every key as a string and return a new string seems a bit off.
> 
> Also, perhaps assert that the new env is different than the incoming environment? I imagine if you get this wrong and it doesn't update anything and your test fails for some weird reason it might be confusing. Or perhaps even worse, you get the key name wrong and your test _still_ passes.
> 
> EDIT:
> I just realized that you are swapping the prefix out, which is why you did that. Maybe that is the way to go. I think there might be some benefit to the rename keys where its more clear which keys are getting renamed but way less sure now. Feel free to ignore if none of this swayed you.

Yeah, the intent was to have a convention that a certain prefix is renamed, in bulk, to cover things like variants of the same env (like SSL, or not, possibly other auth mechanisms or tweaks in the future).  That's why I wanted to avoid having to individually enumerate the keys.  However, the "by function" semantic is definitely compelling, as it is more generic (which is on brand for a test util macro).  Will have to give it a bit more thought.

camsaul commented on modules/drivers/oracle/test/metabase/driver/oracle_test.clj:
```diff
@@ -231,3 +233,13 @@
+    (if (System/getenv "MB_ORACLE_SSL_TEST_SSL")
+      (testing "Oracle with SSL connectivity"
+        (tu/with-env-keys-renamed-by #(str/replace-first % "mb-oracle-ssl-test" "mb-oracle-test")
+          (metabase.query-processor-test.order-by-test/order-by-aggregate-fields-test)))
```
I would generally avoid using the full namespace like this; here's my reasoning:

Using something from a namespace without requiring it creates a race condition in the namespace that uses it. e.g. suppose I have a namespace like

```clj
(ns my-namespace)

(defn my-fn [x]
  (clojure.pprint/pprint x))
```

`(require 'my-namespace)` will work fine 99% of the time, so long as `clojure.pprint` has already been loaded by someone else. However, if you attempt to load `my-namespace` before `clojure.pprint` has been loaded, it will fail because `clojure.pprint` doesn't exist (yet). So you have a race condition when loading this namespace, as it's implicitly dependent on someone else loading things it uses before it gets loaded.

So any time you see a symbol qualified by the entire namespace like that, you can't be sure it's not a situation where it's used without being required _unless_ you jump to the `ns` form to double-check. On the other hand, if you use an alias of some sort, anyone can be sure you've properly required the namespace without having to jump to the `ns` form to look.

I know you *did* require this namespace, but it's nicer if other people won't have to double-check this in the future

jeff303 replied:
Thanks for the explanation.  Fixed now.

camsaul commented on modules/drivers/oracle/test/metabase/driver/oracle_test.clj:
```diff
@@ -12,6 +13,7 @@
             [metabase.models.table :refer [Table]]
             [metabase.query-processor :as qp]
             [metabase.query-processor-test :as qp.test]
+            [metabase.query-processor-test.order-by-test] ; used for one SSL connectivity test
```
You can just alias this as `qp-test.order-by-test` or something like that, it doesn't matter too much

jeff303 replied:
Done.

camsaul commented on modules/drivers/oracle/test/metabase/driver/oracle_test.clj:
```diff
@@ -231,3 +233,13 @@
+      (testing "Oracle with SSL connectivity"
+        (tu/with-env-keys-renamed-by #(str/replace-first % "mb-oracle-ssl-test" "mb-oracle-test")
+          (metabase.query-processor-test.order-by-test/order-by-aggregate-fields-test)))
+      (println (u/colorize 'yellow (format "Skipping %s because %s env var is not set\n"
```
there's a `u/format-color` function you can use that combines these

jeff303 replied:
Fixed in both tests.

camsaul commented on test/metabase/driver/mysql_test.clj:
```diff
@@ -339,13 +337,8 @@
-          (with-redefs [env/env new-e]
-            ;; with the mysql-test vars having been set to the mysql-ssl variants, run some test to verify connectivity
-            (metabase.query-processor-test.string-extracts-test/test-breakout))))
+        (tu/with-env-keys-renamed-by #(str/replace-first % "mb-mysql-ssl-test" "mb-mysql-test")
```
You should put an alias to `with-env-keys-renamed-by` in `metabase.test` and refer to it by `mt/with-env-keys-renamed-by`.

`metabase.test.util` is an ok place for it to actually live, but test utils should be consumed thru `metabase.test`.

Before we had `metabase.test`, there were way too many test util namespaces and it made writing tests hard because you had to remember which util function was in each. So `metabase.test` provides almost all of the common test utils in one place for easy access.

I don't want to undo all that hard work and have people start having to use `metabase.test.util` directly again.

jeff303 replied:
Fixed in both tests.

camsaul commented on test/metabase/driver/mysql_test.clj:
```diff
@@ -339,13 +337,8 @@
-            ;; with the mysql-test vars having been set to the mysql-ssl variants, run some test to verify connectivity
-            (metabase.query-processor-test.string-extracts-test/test-breakout))))
+        (tu/with-env-keys-renamed-by #(str/replace-first % "mb-mysql-ssl-test" "mb-mysql-test")
+          (metabase.query-processor-test.string-extracts-test/test-breakout)))
```
Same thing as before -- alias this in your require and then use the alias here so people know the namespace was required without looking at the `ns` form

jeff303 replied:
Done.

camsaul commented on test/metabase/test/util.clj:
```diff
@@ -794,3 +796,15 @@
+  rename-fn."
+  {:arglists '([rename-fn & body])}
+  [rename-fn & body]
+  `(let [e#      env/env
```
1.  You should add some `testing` context here. It's free and it it's always nice to have more information than less when debugging failing tests. See the impl of `with-temporary-setting-values` for example. I'd do a `data/diff` on `env/env` before and after the call to `map-keys` and use that.
2. You'd make your life a lot easier if you implemented the actual logic here a function and made `with-env-keys-renamed-by` just a thin wrapper around the function. The impl of `discard-setting-changes` is a good example of what I mean.

   Not only is it a lot easier to understand, but it results in smaller code/faster compilation since we're not repeating the same code everywhere the macro is used

jeff303 replied:
In order to capture the full set of renames, to set the testing context, I found it simpler to just build up a map of them (via `reduce-kv`) then use `set/rename-keys` with that.  Also, have moved the bulk of the logic to a fn instead as per your suggestion.

camsaul reviewed (changes requested):
1.  You should alias namespaces you require and use those aliases, so it's clear at a glance that you're not using things without requiring them and causing namespace loading race conditions
2. New test util functions should be "imported" by Potemkin in `metabase.test`, and people using them should use them via `mt/`. Don't use `tu/` or other stuff in new tests -- we don't want to be back in the place where people had to use 6 different test util namespaces to get anything done
3. Your macro should include some testing context to make debugging failing tests a little easier. Maybe you accidentally renamed all the env vars; you shouldn't need to debug the test util to figure that out -- it should tell you

jeff303 committed:
PR feedback:

Move bulk of logic from with-env-keys-renamed-by to new do-with-env-keys-renamed-by fn instead, calling that from macro

Change logic of do-with-env-keys-renamed-by to first build the map of renames, then call set/rename-keys with that (for easier capturing of the testing context information)

Switch to u/format-color in tests

Reference new test macro via metabase.test (which imports it)

Use namespace prefixes for the delegated test calls

camsaul reviewed (approved):
Cool LGTM
